torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+    from triton.runtime.jit import JITFunction
+
+    import torch
+    from torch._subclasses.fake_tensor import FakeTensor
 
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            kwargs = {**kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
+    assert isinstance(kernel, JITFunction)
+
+    # Drop the keys
+    # Replace all SymExprs with a regular value for TTIR generation
+    # Replace all FakeTensor with real tensors
+    # These replacements are needed for triton's type, key and config functions
+    args: List[Any] = []
+    for a in kwargs.values():
+        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool)):
+            args.append(2)
+        elif isinstance(a, FakeTensor):
+            args.append(torch.empty(2, dtype=a.dtype))
+        else:
+            args.append(a)
+
+    specialization = kernel._get_config(*args)
+    constants = {i: arg for i, arg in enumerate(args) if not isinstance(arg, Tensor)}
+    debug = None
+    target = None
+
+    # Build kernel signature -- doesn't include constexpr arguments.
+    signature = {
+        i: kernel._type_of(kernel._key_of(arg))
+        for i, arg in enumerate(args)
+        if i not in kernel.constexprs
+    }
+
+    try:
+        ttir_module = ast_to_ttir(
+            kernel, signature, specialization, constants, debug, target
+        )
+        ops = parse_ttir(str(ttir_module), kwargs)
+        if ops is None:
+            raise Exception("Parsing TTIR failed")
+    except Exception as e:
+        import traceback
+
+        log.debug(
+            "Encountered an exception in identify_mutated_tensors, assuming every input is mutated"
+        )
+        log.debug(
+            "".join(
+                traceback.TracebackException.from_exception(e).format()  # noqa: G001
+            )
+        )
+        return [key for key, value in kwargs.items() if isinstance(value, Tensor)]
+
+    # Name of mutation op to mutated parameter indices
+    MUTATION_OPS = {"tt.store": [0], "tt.atomic_cas": [0], "tt.atomic_rmw": [0]}
+    # Ops that we want to bail out on
+    UNKNOWN_OPS = {"tt.elementwise_inline_asm"}
+
+    stack: List[Union[Param, Intermediate]] = []
+    for op in ops.values():
+        if op.name in UNKNOWN_OPS:
+            log.debug(
+                f"ttir analysis hit an op we do not know how to analyze: {op.name}"  # noqa: G004
+            )
+            return [key for key, value in kwargs.items() if isinstance(value, Tensor)]
+
+        for idx in MUTATION_OPS.get(op.name, []):
+            stack.append(op.args[idx])
+
+    # The following is an iterative DFS algorithm
+    mutated = [False] * len(kwargs)
```

Comment:
Maybe also keep the `visited` set? Otherwise the DFS below can take long for densely connected CFGs (not sure if we should expect those in TTIR, but there are large Triton kernels in the wild, so maybe).